File: src/test/java/com/vjiki/music/mapper/SongMapperTest.java
package com.vjiki.music.mapper;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import com.vjiki.music.dto.SongResponse;
import com.vjiki.music.entity.Song;

class SongMapperTest {

    @Nested
    @DisplayName("SongMapper.toResponse()")
    class ToResponse {

        @Test
        @DisplayName("should map song to response correctly")
        void shouldMapSongToResponseCorrectly() {
            Song song = Song.builder()
                    .id(UUID.randomUUID())
                    .artists(Map.of("default", List.of("Artist Name")))
                    .audioUrls(Map.of("default", "http://audio.com/song.mp3"))
                    .coverUrls(Map.of("default", "http://cover.com/cover.jpg"))
                    .title("Test Song")
                    .active(true)
                    .createdBy("system")
                    .modifiedBy("system")
                    .build();

            SongResponse response = SongMapper.toResponse(song);

            assertThat(response.getId()).isEqualTo(song.getId().toString());
            assertThat(response.getTitle()).isEqualTo("Test Song");
            assertThat(response.getArtist()).isEqualTo("Artist Name");
            assertThat(response.getAudioUrl()).isEqualTo("http://audio.com/song.mp3");
            assertThat(response.getCover()).isEqualTo("http://cover.com/cover.jpg");
        }

        @Test
        @DisplayName("should handle empty artists")
        void shouldHandleEmptyArtists() {
            Song song = Song.builder()
                    .id(UUID.randomUUID())
                    .artists(Map.of())
                    .audioUrls(Map.of("default", "http://audio.com/song.mp3"))
                    .coverUrls(Map.of("default", "http://cover.com/cover.jpg"))
                    .title("Test Song")
                    .active(true)
                    .createdBy("system")
                    .modifiedBy("system")
                    .build();

            SongResponse response = SongMapper.toResponse(song);

            assertThat(response.getArtist()).isNull();
        }

        @Test
        @DisplayName("should handle null artists")
        void shouldHandleNullArtists() {
            Song song = Song.builder()
                    .id(UUID.randomUUID())
                    .artists(null)
                    .audioUrls(Map.of("default", "http://audio.com/song.mp3"))
                    .coverUrls(Map.of("default", "http://cover.com/cover.jpg"))
                    .title("Test Song")
                    .active(true)
                    .createdBy("system")
                    .modifiedBy("system")
                    .build();

            SongResponse response = SongMapper.toResponse(song);

            assertThat(response.getArtist()).isNull();
        }

        @Test
        @DisplayName("should handle missing default in artists")
        void shouldHandleMissingDefaultInArtists() {
            Song song = Song.builder()
                    .id(UUID.randomUUID())
                    .artists(Map.of("other", List.of("Other Artist")))
                    .audioUrls(Map.of("default", "http://audio.com/song.mp3"))
                    .coverUrls(Map.of("default", "http://cover.com/cover.jpg"))
                    .title("Test Song")
                    .active(true)
                    .createdBy("system")
                    .modifiedBy("system")
                    .build();

            SongResponse response = SongMapper.toResponse(song);

            assertThat(response.getArtist()).isNull();
        }

        @Test
        @DisplayName("should handle empty artist list")
        void shouldHandleEmptyArtistList() {
            Song song = Song.builder()
                    .id(UUID.randomUUID())
                    .artists(Map.of("default", List.of()))
                    .audioUrls(Map.of("default", "http://audio.com/song.mp3"))
                    .coverUrls(Map.of("default", "http://cover.com/cover.jpg"))
                    .title("Test Song")
                    .active(true)
                    .createdBy("system")
                    .modifiedBy("system")
                    .build();

            SongResponse response = SongMapper.toResponse(song);

            assertThat(response.getArtist()).isNull();
        }
    }
}

File: bin/test/com/vjiki/music/mapper/SongMapperTest.kt
package com.vjiki.music.mapper

import com.vjiki.music.entity.Song
import com.vjiki.music.mapper.SongMapper.toResponse
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import java.util.*

class SongMapperTest : DescribeSpec({

    describe("Song.toResponse()") {
        it("should map song to response correctly") {
            val song = Song(
                id = UUID.randomUUID(),
                artists = mapOf("default" to listOf("Artist Name")),
                audioUrls = mapOf("default" to "http://audio.com/song.mp3"),
                coverUrls = mapOf("default" to "http://cover.com/cover.jpg"),
                title = "Test Song",
                active = true,
                createdBy = "system",
                modifiedBy = "system"
            )

            val response = song.toResponse()

            response.id shouldBe song.id.toString()
            response.title shouldBe "Test Song"
            response.artist shouldBe "Artist Name"
            response.audioUrl shouldBe "http://audio.com/song.mp3"
            response.cover shouldBe "http://cover.com/cover.jpg"
        }

        it("should handle null artists") {
            val song = Song(
                id = UUID.randomUUID(),
                artists = emptyMap(),
                audioUrls = mapOf("default" to "http://audio.com/song.mp3"),
                coverUrls = mapOf("default" to "http://cover.com/cover.jpg"),
                title = "Test Song",
                active = true,
                createdBy = "system",
                modifiedBy = "system"
            )

            val response = song.toResponse()

            response.artist shouldBe null
        }

        it("should handle missing default in artists") {
            val song = Song(
                id = UUID.randomUUID(),
                artists = mapOf("other" to listOf("Other Artist")),
                audioUrls = mapOf("default" to "http://audio.com/song.mp3"),
                coverUrls = mapOf("default" to "http://cover.com/cover.jpg"),
                title = "Test Song",
                active = true,
                createdBy = "system",
                modifiedBy = "system"
            )

            val response = song.toResponse()

            response.artist shouldBe null
        }

        it("should handle empty artist list") {
            val song = Song(
                id = UUID.randomUUID(),
                artists = mapOf("default" to emptyList()),
                audioUrls = mapOf("default" to "http://audio.com/song.mp3"),
                coverUrls = mapOf("default" to "http://cover.com/cover.jpg"),
                title = "Test Song",
                active = true,
                createdBy = "system",
                modifiedBy = "system"
            )

            val response = song.toResponse()

            response.artist shouldBe null
        }
    }
})

